src/geometry.cpp

```diff
@@ -107,5 +111,87 @@ namespace geometry {
 
 	}
 
+	std::vector<Cell> polygonOutlineCells(const std::vector<Cell>& polygon, int size_x) {	  
+	  // The output vector of cells
+	  std::vector<Cell> out;	  
+
+	  // If the input is less than or equal to one element, return the input
+	  if (polygon.size() <= 1) {
+	  	return polygon;
+	  }
+
+	  // Iterator to the next vertex
+	  auto next_vrtx = polygon.begin();
+
+	  for (const auto &vertex:polygon) {
+	  	// If the next vertex is not the end iterator, advance the next_vertex to point to the next iterator
+	  	if (next_vrtx != polygon.end()) {
+	  		std::advance(next_vrtx,1);
+	  	} 
+	  	// After advancing, if the iterator points to the end, go back to the beginning, essentially closing the polygon
+	  	if (next_vrtx == polygon.end()) {
+	  		next_vrtx = polygon.begin();
+	  	}
+
+	  	// Generate the pixels from one vertex to the next
+	  	auto temp = raytrace(vertex, (*next_vrtx), size_x);
+
+	  	// Remove the last element of the vector because the next vertex will use it as a starting point i.e. don't duplicate pixels
+	  	temp.pop_back();
+
+	  	// Push the pixels into the out vector
+	  	for (const auto &pixel:temp) {
+	  		out.push_back(pixel);
+	  	}
+
+	  }
+	  return out;
+	}	
+
+	std::vector<Cell> convexFillCells(const int size_x, const std::vector<Cell>& polygon) {
+	  // we need a minimum polygon of a triangle (Should this return the polygon?)
+	  if (polygon.size() < 3) return polygon;
+
+	  const auto polygon_outline = polygonOutlineCells(polygon, size_x);
+
+	  std::unordered_map<int,std::pair<int,int>> x_and_extrema;
+
+	  //TODO: turn the nested if else to else ifs
+
+	  // Find all the x values and extrema
+	  for (const auto pixel:polygon_outline) {
+	  	// Check if the x value for the pixel is in x_and_extrema
+	  	if (x_and_extrema.find(pixel.x) == x_and_extrema.end()) {
+	  		// if it isnt in x_and_extrema push the x in and with the y value as both the min and the max
+	  		x_and_extrema[pixel.x] = {pixel.y, pixel.y};
+	  	} else {
+	  		// if it is in x_and_extrema, check the y value. If the y value is lower than the minima,
+	  		// replace the first value, else if it is greater than the maxima, replace the second 
+	  		// value, else do nothing
+	  		if (pixel.y < ((x_and_extrema.at(pixel.x)).first)) {
+	  			(x_and_extrema.at(pixel.x)).first = pixel.y;
+	  		} else if (pixel.y > ((x_and_extrema.at(pixel.x)).second)) {
+	  			(x_and_extrema.at(pixel.x)).second = pixel.y;
+	  		}
+	  	}
+	  }
+
+	  std::vector<Cell> out;
```

Comment:
https://www.reddit.com/r/cpp/comments/9prei7/stl_algorithms_11_stdgenerate_vs_stdgenerate_n/

FYI, if we end up using this library a lot, we might end up trying to figure out how to preallocate, given the cpu savings